sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,591 @@
+                    verb: "create directory",
+                    path: temp_dir,
+                    err,
+                });
+            }
+        }
+        let (file, temp_path) = tokio::task::spawn_blocking(move || {
+            NamedUtf8TempFile::new_in(&temp_dir).map_err(|err| Error::File {
+                verb: "create temporary file in",
+                path: temp_dir,
+                err,
+            })
+        })
+        .await??
+        .into_parts();
+        let file =
+            NamedUtf8TempFile::from_parts(File::from_std(file), temp_path);
+
+        let (mut file, hasher) = stream
+            .try_fold(
+                (file, Sha256::new()),
+                |(mut file, mut hasher), chunk| async move {
+                    hasher.update(&chunk);
+                    match file.as_file_mut().write_all(chunk.as_ref()).await {
+                        Ok(()) => Ok((file, hasher)),
+                        Err(err) => Err(Error::File {
+                            verb: "write to",
+                            path: file.path().to_owned(),
+                            err,
+                        }),
+                    }
+                },
+            )
+            .await?;
+        file.as_file_mut().sync_data().await.map_err(|err| Error::File {
+            verb: "sync",
+            path: final_path.clone(),
+            err,
+        })?;
+
+        let digest = hasher.finalize();
+        if digest.as_slice() != sha256.as_ref() {
+            return Err(Error::HashMismatch {
+                expected: sha256,
+                actual: ArtifactHash(digest.into()),
+            });
+        }
+
+        let moved_final_path = final_path.clone();
+        tokio::task::spawn_blocking(move || {
+            file.persist(&moved_final_path).map_err(|err| Error::File {
+                verb: "rename temporary file to",
+                path: moved_final_path,
+                err: err.error,
+            })
+        })
+        .await??;
+        info!(
+            &self.log,
+            "Wrote artifact";
+            "sha256" => &sha256.to_string(),
+            "path" => final_path.as_str(),
+        );
+        Ok(())
+    }
+
+    /// PUT operation (served by Sled Agent API) which takes a [`StreamingBody`]
+    pub(crate) async fn put_body(
+        &self,
+        sha256: ArtifactHash,
+        body: StreamingBody,
+    ) -> Result<(), Error> {
+        self.put(sha256, body.into_stream().map_err(Error::Body)).await
+    }
+
+    /// POST operation (served by Sled Agent API)
+    pub(crate) async fn copy_from_depot(
+        &self,
+        sha256: ArtifactHash,
+        address: SocketAddrV6,
+    ) -> Result<(), Error> {
+        static CLIENT: LazyLock<reqwest::Client> = LazyLock::new(|| {
+            reqwest::ClientBuilder::new()
+                .connect_timeout(Duration::from_secs(15))
+                .read_timeout(Duration::from_secs(15))
+                .build()
+                .unwrap()
```

Comment:
Maybe although I'm not sure what I'd say other than "this unwrap is the same as what [every other Progenitor client does](https://github.com/oxidecomputer/progenitor/blob/v0.8.0/progenitor-impl/src/lib.rs#L440-L456)". It's not actually 100%, because the OpenSSL backend can fail while reading the system certificate store, and we'd unwrap that error. But we also clearly do not want to continue in that case either.

It's a little safer if the client is hanging off of `self` instead of being run the first time `copy_from_depot` is called though.